Automattic/jetpack
# [CRM] Make Automations Workflow endpoints functional

This PR will update the "Automation Workflows endpoints" that was previously added with dummy data, to being fully functional now. It aims to do so by:

* Complete the `[GET] /wp-json/jetpack-crm/v4/automation/workflows` endpoint
  * Params: `(bool) active`, `(string) category`, `(int) page`, `(int) per_page`, `(int) offset`.
* Complete the `[GET] /wp-json/jetpack-crm/v4/automation/workflows/{id}` endpoint
  * This only accepts the ID in the URL. 
* Complete the `[PUT] /wp-json/jetpack-crm/v4/automation/workflows/{id}` endpoint
  * Params: `(string) name`, `(string) description`, `(bool) active`, `(string) category`, `(string|int) initial_step`, `(object) steps`.
    * The "Steps" param is registered as an "object" because WP REST assumes "array" is a numeric array but we want to allow both strings and integers as the key. We also require a "slug" to be present as a simple verification that we're getting the correct format.

Fixes https://github.com/Automattic/zero-bs-crm/issues/3226

## Proposed changes:

* Rename controller from `REST_Automation_Controller` to `REST_Automation_Workflows_Controller` since, theoretically, if we add other endpoints than workflows, then it makes sense to separate them to improve readability and maintainability.
    * I've also renamed the `REST_Automation_Controller_Test` test class to `REST_Automation_Workflows_Controller_Test` to reflect the above change.
* Renamed methods inside the workflow controller to match default WordPress naming (`get_item`, `get_items`, etc).
* Added pagination to `Workflow_Repository::find_by`
* Changed the singular workflow paths to follow RESTful naming (`/jetpack-crm/v4/automation/workflow` => `/jetpack-crm/v4/automation/workflows/{id}`)
* Complete the workflow endpoints with dummy data:
    * Complete the `[GET] /wp-json/jetpack-crm/v4/automation/workflows` endpoint
    * Complete the `[GET] /wp-json/jetpack-crm/v4/automation/workflows/{id}` endpoint
    * Complete the `[PUT] /wp-json/jetpack-crm/v4/automation/workflows/{id}` endpoint
* Updated tests to use the workflow repository instead of relying on the dummy data that existed in the endpoint skeleton versions.

**Side-note:**

I've tried to keep a link between the old controller and test files, but I cannot for the life of me get it to work right now. So to save time while nearing the ETA, I've just chosen to accept it as is.

### Other information:


## Jetpack product discussion



* pbhBOz-3ss-p2
* pbhBOz-2ES-p2
* pbhBOz-3uX-p2

## Does this pull request change what data or activity we track or use?



No.

## Testing instructions:






### Automated tests

All of the tests should obviously pass, but looking at the following automated tests should suffice:

* `REST_Authentication_Test`: Updated to include our new endpoints.
* `REST_Automation_Workflows_Controller_Test`: Updated to use real integration tests + added new tests to verify additional expected behaviour like 404 response(s).
* `Workflow_Repository_Test`: Updated to test limit/offset arguments when fetching workflows in bulk.

### Manual testing

**Manual testing prep work**

1. Make sure the following feature flags are enabled on your site:

```php
add_filter( 'jetpack_crm_feature_flag_api_v4', '__return_true' );
add_filter( 'jetpack_crm_feature_flag_automations', '__return_true' );
```

2. Modify `REST_Automation_Workflows_Controller::get_item_permissions_check` to return `true` at the top.

_We do this to make it easier to call the endpoints. This should allow you to do it from any client you want (e.g. Postman) without having to worry about WP Logged in cookies, nonce, etc._

3. Make sure you have a workflow in your database. You can use the below snippet to add one.

```php
$workflow = new \Automattic\Jetpack\CRM\Automation\Automation_Workflow(
	[
		'name'         => 'New contact: Set status to "Customer"',
		'description'  => 'This automation will change the status of a contact to "Customer" when they are created.',
		'category'     => 'contact',
		'active'       => true,
		'triggers'     => [
			\Automattic\Jetpack\CRM\Automation\Triggers\Contact_Created::get_slug(),
		],
		'initial_step' => 1,
		'steps'        => [
			1 => [
				'slug'       => \Automattic\Jetpack\CRM\Automation\Actions\Update_Contact_Status::get_slug(),
				'attributes' => [
					'new_status' => 'Customer',
				],
				'next_step'  => null,
			],
		],
	]
);

$repo = new \Automattic\Jetpack\CRM\Automation\Workflow\Workflow_Repository();
$workflow = $repo->persist(  $workflow );
```

**GET All Workflows**

* Method: `GET`
* Path: `/wp-json/jetpack-crm/v4/automation/workflows`
* Payload: N/A

**GET Workflow**

* Method: `GET`
* Path: `/wp-json/jetpack-crm/v4/automation/workflows/{id}`
* Payload: N/A

**Update Workflow**

* Method: `PUT` (or `POST`, both works)
* Path: `/wp-json/jetpack-crm/v4/automation/workflows/{id}`
* Payload: `{ "name": "My updated name" }`
